compiler/llvm/clangUtil.cpp

```diff
@@ -4251,13 +4546,34 @@ static void makeBinaryLLVMForHIP(const std::string& artifactFilename,
   mysystem(bundlerCmd.c_str(), ".out file to fatbin file");
 }
 
-void makeBinaryLLVM(void) {
+static void saveIrToBcFileIfNeeded(const std::string& filename) {
+  GenInfo* info = gGenInfo;
+  if (saveCDir[0] != '\0') {
+    std::error_code tmpErr;
+    // Save the generated LLVM IR to the given file
+#if HAVE_LLVM_VER >= 120
+    ToolOutputFile output (filename.c_str(), tmpErr, sys::fs::OF_None);
+#else
+    ToolOutputFile output (filename.c_str(), tmpErr, sys::fs::F_None);
+#endif
+    if (tmpErr) {
+      USR_FATAL("Could not open output file %s", filename.c_str());
+    }
+    WriteBitcodeToFile(*info->module, output.os());
+    output.keep();
+    output.os().flush();
+  }
+}
 
+void makeBinaryLLVM(void) {
   GenInfo* info = gGenInfo;
   INT_ASSERT(info);
   ClangInfo* clangInfo = info->clangInfo;
   INT_ASSERT(clangInfo);
 
+  // the per-function optimization pipeline is no longer needed
+  cleanupFunctionOptManagers();
```

Comment:
This is also called in finishCodegenLLVM, is this not redundant?

Reply:
It is redundant, but the idea is that we can clean up the memory from these early in normal operation. There might be code paths that don't run this function but do run finishCodegenLLVM.